### PR TITLE
DM-25894: Extract stars on, or slightly beyond, exposure edge

### PR DESCRIPTION
**Pull request by MorganSchmitz**
[Jira ticket](https://jira.lsstcorp.org/browse/DM-25894)

[Jenkins run #33446](https://ci.lsst.codes/blue/organizations/jenkins/stack-os-matrix/detail/stack-os-matrix/33446/pipeline)

### CODE REVIEW

File: python/lsst/pipe/tasks/processBrightStars.py [hunk 1]
```diff
@@ -127,6 +128,12 @@ class ProcessBrightStarsConfig(pipeBase.PipelineTaskConfig,
             " annular flux.",
         default=('BAD', 'CR', 'CROSSTALK', 'EDGE', 'NO_DATA', 'SAT', 'SUSPECT', 'UNMASKEDNAN')
     )
+    minPixelsWithinFrame = pexConfig.Field(
+        dtype=int,
+        doc="When a bright star is beyond exposure boundary, what is the minimal number of pixels that still "
```
Comment (sr525) [inline python/lsst/pipe/tasks/processBrightStars.py:133]: I would suggest rewording to something like: The minimum number of pixels that must fall within the stamp boundary for the bright star to be saved when it is beyond the exposure boundary.

File: python/lsst/pipe/tasks/processBrightStars.py [hunk 2]
```diff
@@ -207,8 +214,12 @@ def extractStamps(self, inputExposure, refObjLoader=None):
         GMags = []
         ids = []
         wcs = inputExposure.getWcs()
-        # select stars within input exposure from refcat
-        withinCalexp = refObjLoader.loadPixelBox(inputExposure.getBBox(), wcs, filterName="phot_g_mean")
+        bmp = inputExposure.mask.getMaskPlaneDict()
```
Comment (sr525) [inline python/lsst/pipe/tasks/processBrightStars.py:217]: I would prefer a more descriptive name than bmp.

File: python/lsst/pipe/tasks/processBrightStars.py [hunk 3]
```diff
@@ -207,8 +214,12 @@ def extractStamps(self, inputExposure, refObjLoader=None):
...
+        # select stars within, or close enough to input exposure from refcat
+        inputExpBBox = inputExposure.getBBox()
+        dilatationExtent = geom.Extent2I(np.array(self.config.stampSize) - self.config.minPixelsWithinFrame)
+        withinCalexp = refObjLoader.loadPixelBox(inputExpBBox.dilatedBy(dilatationExtent), wcs,
+                                                 filterName="phot_g_mean")
```
Comment (sr525) [inline python/lsst/pipe/tasks/processBrightStars.py:222]: Is the filter always going to be phot_g_mean?

Comment (MorganSchmitz) [reply python/lsst/pipe/tasks/processBrightStars.py:222]: For now, yes. The refcat used will always be Gaia, and the band always (Gaia) G.

This will only change with DM-25175, and I was planning on adjusting this when we get there. I added a TODO comment referencing the tickets for now.

File: python/lsst/pipe/tasks/processBrightStars.py [hunk 4]
```diff
@@ -221,15 +233,38 @@ def extractStamps(self, inputExposure, refObjLoader=None):
...
-                GMags.append(allGMags[j])
-                ids.append(allIds[j])
+            try:
+                starIm = inputExposure.getCutout(sp, geom.Extent2I(self.config.stampSize))
+            except InvalidParameterError:
+                # star is beyond boundary
+                bboxCorner = np.array(cpix) - np.array(self.config.stampSize)/2
+                bboxCorner = geom.Point2I(bboxCorner)
+                # compute bbox as it would be otherwise
+                idealBBox = geom.Box2I(geom.Point2I(bboxCorner), geom.Extent2I(self.config.stampSize))
```
Comment (sr525) [inline python/lsst/pipe/tasks/processBrightStars.py:243]: You've already cast it as a Point2I

File: python/lsst/pipe/tasks/processBrightStars.py [hunk 5]
```diff
@@ -221,15 +233,38 @@ def extractStamps(self, inputExposure, refObjLoader=None):
...
-                    and cpix[0] < inputExposure.getDimensions()[0] - self.config.stampSize[0]/2
-                    and cpix[1] >= self.config.stampSize[1]/2
-                    and cpix[1] < inputExposure.getDimensions()[1] - self.config.stampSize[1]/2):
-                starIms.append(inputExposure.getCutout(sp, geom.Extent2I(self.config.stampSize)))
-                pixCenters.append(cpix)
-                GMags.append(allGMags[j])
-                ids.append(allIds[j])
+            try:
+                starIm = inputExposure.getCutout(sp, geom.Extent2I(self.config.stampSize))
+            except InvalidParameterError:
```
Comment (sr525) [inline python/lsst/pipe/tasks/processBrightStars.py:238]: I have slacked you a suggestion for how to replace this section.

**Comment by MorganSchmitz** [comment]
Thanks a lot for your suggestion. I went with a mix between your version and mine, but it is a lot easier to read that way!

File: python/lsst/pipe/tasks/processBrightStars.py [hunk 6]
```diff
@@ -233,32 +245,49 @@ def extractStamps(self, inputExposure, refObjLoader=None):
...
+                detThreshold = afwDetect.Threshold(starIm.mask.getPlaneBitMask("DETECTED"),
+                                                   afwDetect.Threshold.BITMASK)
+                omask = afwDetect.FootprintSet(starIm.mask, detThreshold)
+                allFootprints = omask.getFootprints()
+                otherFootprints = []
+                for fs in allFootprints:
+                    if not fs.contains(geom.Point2I(cpix)):
+                        otherFootprints.append(fs)
+                nbMatchingFootprints = len(allFootprints) - len(otherFootprints)
+                if not nbMatchingFootprints == 1:
```
Comment (sr525) [inline python/lsst/pipe/tasks/processBrightStars.py:282]: Is this the only criteria that can trigger this condition?
